### ISSUE TITLE
selfHeal true skips auto-sync

### ISSUE THREAD

**Issue by dontreboot** [argoproj/argo-cd#4101]
If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a
question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

With selfHeal enabled, ArgoCD still skips auto-sync

**To Reproduce**
This is the application manifest I have when a sync fails, ArgoCD skips auto-sync
```
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  labels:
    app.kubernetes.io/instance: argocd
  name: someapp
  namespace: argocd
spec:
  destination:
    namespace: someapp
    server: 'https://kubernetes.default.svc'
  project: someapp
  source:
    path: x
    repoURL: x
    targetRevision: master
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
```
**Expected behavior**

I was under the impression that with `selfHeal`eanabled ArgoCD would re-attempt sync again after default self-heal-timeout-seconds. Maybe I was wrong but I wanted to make sure what the right exception for selfHeal was.

**Screenshots**

**Version**

```shell
argocd: v1.6.2+3d1f37b.dirty
  BuildDate: 2020-08-01T06:59:53Z
  GitCommit: 3d1f37b0c53f4c75864dc7339e2831c6e6a947e0
  GitTreeState: dirty
  GoVersion: go1.14.6
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v1.6.1+159674e
  BuildDate: 2020-06-19T00:41:05Z
  GitCommit: 159674ee844a378fb98fe297006bf7b83a6e32d2
  GitTreeState: clean
  GoVersion: go1.14.1
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: {Version:kustomize/v3.6.1 GitCommit:c97fa946d576eb6ed559f17f2ac43b3b5a8d5dbd BuildDate:2020-05-27T20:47:35Z GoOs:linux GoArch:amd64}
  Helm Version: version.BuildInfo{Version:"v3.2.0", GitCommit:"e11b7ce3b12db2941e90399e874513fbd24bcb71", GitTreeState:"clean", GoVersion:"go1.13.10"}
  Kubectl Version: v1.14.0
```

**Logs**

```
  ""log"": ""time=\""2020-08-05T23:12:29Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:15:00Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:16:04Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:16:04Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:17:04Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:18:04Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:19:06Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:19:06Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:20:05Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:22:59Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:22:59Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:23:59Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:25:01Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:27:06Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:27:06Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:28:05Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:29:07Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:30:11Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:30:12Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:30:57Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:31:59Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:33:01Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:33:27Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:34:03Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:36:27Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:37:09Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:38:59Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",
  ""log"": ""time=\""2020-08-05T23:39:28Z\"" level=warning msg=\""Skipping auto-sync: failed previous sync attempt to 433a1cdb1a67efa86c8c1a48d7137c05f06891f3\"" application=someapp\n"",```

**Comment by jessesuen** [comment]
This is expected behavior. When an auto-sync fails, we do not attempt to auto-heal because the healing process would miss out from hooks that were supposed to happen during the auto-sync. 

There is a separate feature to retry automated syncs.

**Comment by alexmt** [comment]
Failed sync retries feature had been implemented in v1.7: https://github.com/argoproj/argo-cd/pull/3997
